File: eval/sofa/alert_ids.py
"""Canonical alert-id algorithm (matches Java AlertIds / UUID.nameUUIDFromBytes)."""

from __future__ import annotations

import hashlib
import uuid


def name_uuid_from_bytes(name: str) -> uuid.UUID:
    digest = bytearray(hashlib.md5(name.encode("utf-8")).digest())
    digest[6] = (digest[6] & 0x0F) | 0x30  # version 3
    digest[8] = (digest[8] & 0x3F) | 0x80  # IETF variant
    return uuid.UUID(bytes=bytes(digest))


def alert_id(
    patient_id: str,
    encounter_id: str | None,
    indicator: str,
    score: int | None,
    event_time_ms: int,
    version: str,
) -> str:
    raw = (
        f"{patient_id}|{encounter_id or ''}|{indicator}|{score}|{event_time_ms}|{version}"
    )
    return "alert-" + str(name_uuid_from_bytes(raw))

File: streaming/flink-jobs/sofa/src/main/java/com/curie/sofa/model/AlertIds.java
package com.curie.sofa.model;

import java.nio.charset.StandardCharsets;
import java.util.UUID;

/** Canonical alert-id algorithm shared by SOFA and AKI Flink paths. */
public final class AlertIds {
  private AlertIds() {}

  public static String of(
      String patientId,
      String encounterId,
      String indicator,
      Integer score,
      long eventTimeMs,
      String version) {
    String raw =
        String.valueOf(patientId)
            + "|"
            + (encounterId == null ? "" : encounterId)
            + "|"
            + (indicator == null ? "sofa-deterioration" : indicator)
            + "|"
            + score
            + "|"
            + eventTimeMs
            + "|"
            + (version == null ? "" : version);
    return "alert-" + UUID.nameUUIDFromBytes(raw.getBytes(StandardCharsets.UTF_8));
  }
}
